test/API/H5_api_dataset_test.c

```diff
@@ -10251,6 +10265,9 @@ test_dataset_vlen_io(void)
             if ((H5Dwrite(dset_int, vlen_int, space_id, space_id, H5P_DEFAULT, (const void *)wbuf)) < 0)
                 PART_TEST_ERROR(rw_point_selection);
 
+            if (H5Dflush(dset_int) < 0)
```

Comment:
Hmm I'm surprised adding a flush here would fix the problem since I'd assume that would just move the segfault to here instead. Plus, `H5Dflush` is mostly used for SWMR and I wouldn't think it would be needed here. Since the native VOL was passing without this call, I wonder if this is either a problem in the cache VOL or a problem in the VOL interface instead.

Reply:
The segfault seems to be due to the cache vol still using the dynamically allocated heap buffers when they're freed after the reads/writes. As it turns out, this doesn't resolve the issue either - it cropped up again in the CI [here](https://github.com/mattjala/hdf5/actions/runs/8295314503). The cache VOL docs say that H5Dwrite/H5Dclose should be semi-blocking, so this may be an issue on the cache VOL.